### PR TITLE
Introduce interval to SQL

### PR DESCRIPTION
**Pull request by ImeevMA**
This patch-set introduces basic INTERVAL support in SQL.

### CODE REVIEW

**Comment by ImeevMA** [comment]
Hi @Gerold103! At first I planned not to include the last two patches in this PR, but in the end I changed my mind. I hope this didn't bother you.

File: src/box/sql/mem.c [hunk 1]
```diff
@@ -2034,15 +2039,86 @@ mem_add(const struct Mem *left, const struct Mem *right, struct Mem *result)
...
+		return -1;
+	}
+	if (mem_is_datetime(right))
+		return mem_add_dt(right, left, result);
+	struct interval res = left->u.itv;
+	if (interval_interval_add(&res, &right->u.itv) != 0)
+		return -1;
+	mem_set_interval(result, &res);
+	return 0;
+}
```
Comment (Gerold103) [inline src/box/sql/mem.c:2086]: You can simplify this all a bit (same for the subtraction):
```Diff
diff --git a/src/box/sql/mem.c b/src/box/sql/mem.c
index bc9f302cc..063ea69ab 100644
--- a/src/box/sql/mem.c
+++ b/src/box/sql/mem.c
@@ -2053,11 +2053,8 @@ mem_add_dt(const struct Mem *left, const struct Mem *right, struct Mem *result)
 			 "interval");
 		return -1;
 	}
-	struct datetime res = left->u.dt;
-	if (datetime_increment_by(&res, 1, &right->u.itv) != 0)
-		return -1;
-	mem_set_datetime(result, &res);
-	return 0;
+	mem_set_datetime(result, &left->u.dt);
+	return datetime_increment_by(&result->u.dt, 1, &right->u.itv);
 }
 
 /**
@@ -2070,19 +2067,15 @@ static int
 mem_add_itv(const struct Mem *left, const struct Mem *right, struct Mem *result)
 {
 	assert(mem_is_interval(left) && !mem_is_metatype(left));
-	if (!(mem_is_interval(right) || mem_is_datetime(right)) ||
-	    mem_is_metatype(right)) {
+	if (mem_is_datetime(right))
+		return mem_add_dt(right, left, result);
+	if (!mem_is_interval(right) || mem_is_metatype(right)) {
 		diag_set(ClientError, ER_SQL_TYPE_MISMATCH, mem_str(right),
 			 "datetime or interval");
 		return -1;
 	}
-	if (mem_is_datetime(right))
-		return mem_add_dt(right, left, result);
-	struct interval res = left->u.itv;
-	if (interval_interval_add(&res, &right->u.itv) != 0)
-		return -1;
-	mem_set_interval(result, &res);
-	return 0;
+	mem_set_interval(result, &left->u.itv);
+	return interval_interval_add(&result->u.itv, &right->u.itv);
 }
```

File: src/box/sql/mem.c [hunk 2]
```diff
@@ -3390,6 +3457,16 @@ port_c_get_vdbemem(struct port *base, uint32_t *size)
 				}
 				val[i].type = MEM_TYPE_DATETIME;
 				break;
+			} else if (type == MP_INTERVAL) {
+				struct interval *itv = &val[i].u.itv;
+				if (interval_unpack(&data, itv) == 0) {
```
[the remark this answers is not in the document]
Comment (ImeevMA) [reply src/box/sql/mem.c]: Thanks, fixed.

File: src/box/sql/mem.c (hunk 1 above)
Comment (ImeevMA) [reply src/box/sql/mem.c]: Thank you for the diff. I applied, but I added one more condition in one of `if`s:
```
+	if (mem_is_datetime(right) && !(mem_is_metatype(right))
+		return mem_add_dt(right, left, result);
```

File: src/box/sql/mem.c [hunk 3]
```diff
@@ -2043,14 +2155,86 @@ mem_sub(const struct Mem *left, const struct Mem *right, struct Mem *result)
...
+	if (!mem_is_interval(right) || mem_is_metatype(right)) {
+		diag_set(ClientError, ER_SQL_TYPE_MISMATCH, mem_str(right),
+			 "interval");
+		return -1;
+	}
+	struct interval res = left->u.itv;
+	if (interval_interval_sub(&res, &right->u.itv) != 0)
+		return -1;
+	mem_set_interval(result, &res);
+	return 0;
```
Comment (Gerold103) [inline src/box/sql/mem.c:2207]: I suggest to update the subtractions too, like the additions. Not to create a temporary interval nor datetime on the stack.

**Review by Gerold103 — CHANGES_REQUESTED** [review]

File: src/box/sql/mem.c (hunk 1 above)
Comment (Gerold103) [reply src/box/sql/mem.c]: In the diff I didn't change the subtractions. I only mentioned them in the comment. Lets do the same for them.

File: src/box/sql/mem.c (hunk 3 above)
Comment (ImeevMA) [reply src/box/sql/mem.c]: Fixed, thank you. Sorry, I forgot that we have addition and subtraction.

**Review by Gerold103 — APPROVED** [review]
Thank for the fixes, looks good now!

**Review by kyukhin — APPROVED** [review]

**Pull request merged**